### PR TITLE
feat: add altitude C chord warning sounds

### LINKED ISSUE #1227: [BUG] altitude alert
**Mod Version**
0.3.1.

**Describe the bug**

If you are reaching or leaving your preselect or managed altitude there should be an alert sound.

**To Reproduce**

1.set altitude
2.climb
3.reach altitude

**Expected behavior**

**Actual behavior**

**References**

- (redacted by admin)

**Additional context**

Was this working before/when did the issue start occurring?

Is this a problem in the vanilla unmodded game?

Discord username (if different from GitHub):

### PR DESCRIPTION
**Pull request by Kimbyeoungjang**
Fixes #1227
Closes #2136

## Summary of Changes

using @ilyeshammadi code, add c chord warning sounds

## Screenshots (if necessary)

https://drive.google.com/file/d/1kO6RZTWC-kokUShuDRNvxCrDxzQz6Ywd/view?usp=sharing
you can check vid

## References

actually this image is wrong(not 250ft, it is 200ft.)
image

image

## Additional context

Discord username (if different from GitHub):

## Testing instructions

this pr is basically working like upper document.

so when you test this, spawn on ruwnay, takeoff, and manually flight and check the document.

if you fly manually, when you reach a (seclected altitude - 750)ft, you can hear short cchord alert
when you descend on document scenario, you can here continuous cchord alert.

if you fly with autopilot, only working continuous cchord. when you engage autopilot, idle thurst, deploy speed brake, gear down etc... you reduce the speed, aircraft will stall, you can hear some sitation continuous cchord alert.

just flight as you want with manually, with select altitude, and check the behavior to use that document

also, you can stop the alert with clicking Master Warning or change selected altitude

## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

### CODE REVIEW

**Comment by Kimbyeoungjang** [comment]
ignore some dirty commit, it just update branch from master.

**Comment by Kimbyeoungjang** [comment]
this using my warning logic. so some of my pr is merged(i wish), need to fix my other pr.

**Comment by Kimbyeoungjang** [comment]
ready to test i think :)

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js [hunk 1]
```diff
@@ -56,6 +56,16 @@ class A32NX_FWC {
...
+
+        // master warning & caution buttons
+        this.warningPressed = false;
+        this.cautionPressed = false;
+
+        // altitude warning
+        this.previousTargetAltitude = NaN;
+        this._wasBellowThreshold = false;
+        this._wasAboveThreshold = false;
+        this._wasInRange = false;
```
Comment (vkrizan) [inline A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js:68]: I was wondering whether saving previous delta or use of a state machine would reduce number of these variables.

Then it could be something like:
* `if prevDelta > 750 && currDelta <= 750` -> `ALT_DEVIATION_SHORT` (1.5s)
* `if prevDelta <= 250 && currDelta > 250` -> `ALT_DEVIATION` (continuous)
* `if 750 >= prevDelta > 250 && currDelta > 750` -> `ALT_DEVIATION` (continuous)

Of course there would have to be conditions incorporated in case the warring is inhibited.
The continuous sound/boolean would get precedence over short one.

With this I see one problem of sounding again when the button is pushed between 750ft and 250ft when increasing deviation from selected altitude after passing deviation over 750 ft. I'm not sure if it is a problem or a feature.

If the continuous cchord should stop when reaching 750ft, then it could reset on the same condition as for the short (1.5) cchord.

(pardon my pseudo code)

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js [hunk 2]
```diff
@@ -286,4 +309,80 @@ class A32NX_FWC {
         this.ldgMemo = showInApproach || invalidRadioMemo || this.flightPhase === 8 || this.flightPhase === 7;
         SimVar.SetSimVarValue("L:A32NX_FWC_LDGMEMO", "Bool", this.ldgMemo);
     }
+
+    altitudeWarning(_deltaTime) {
+        const indicatedAltitude = Simplane.getAltitude();
+        SimVar.SetSimVarValue("L:A32NX_ALT_DEVIATION_SHORT", "Bool", false);
+        SimVar.SetSimVarValue("L:A32NX_ALT_DEVIATION", "Bool", false);
```
Comment (vkrizan) [inline A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js:316]: Although I'm not fully aware how the button and short sound booleans work, the use of transitions and states could prevent an excessive reset of these deviation booleans on every frame update.

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js [hunk 3]
```diff
@@ -286,4 +309,80 @@ class A32NX_FWC {
...
+                SimVar.SetSimVarValue("L:A32NX_ALT_DEVIATION", "Bool", true);
+            } else if (this._wasInRange) {
+                if (SimVar.GetSimVarValue("L:XMLVAR_Autopilot_1_Status", "Bool") === 0 && SimVar.GetSimVarValue("L:XMLVAR_Autopilot_2_Status", "Bool") === 0) {
+                    SimVar.SetSimVarValue("L:A32NX_ALT_DEVIATION_SHORT", "Bool", true);
+                }
+            }
+        }
+    }
+
+    getAutopilotMode() {
```
Comment (vkrizan) [inline A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js:380]: To better map the semantics, I'd suggest to name it something like `hasAltitudeConstraint`.

**Review by vkrizan — COMMENTED** [review]
Awesome stuff. Thank you!

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js (hunk 1 above)
Comment (Kimbyeoungjang) [reply A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js:68]: thanks for your feedback. i try to reduce simvars and change some code on that.
i just using pfd code(copy & paste) but try to re-build that :)

Comment (Kimbyeoungjang) [reply A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js:68]: actually this code is using on PFD so i think it work well.

so i try to reduce simvar only :)
i think this will working well but if there is a error on qa test, then fix that!

**Comment by Kimbyeoungjang** [comment]
finish!

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js [hunk 4]
```diff
@@ -286,4 +309,81 @@ class A32NX_FWC {
         this.ldgMemo = showInApproach || invalidRadioMemo || this.flightPhase === 8 || this.flightPhase === 7;
         SimVar.SetSimVarValue("L:A32NX_FWC_LDGMEMO", "Bool", this.ldgMemo);
     }
+
+    altitudeWarning(_deltaTime) {
+        const indicatedAltitude = Simplane.getAltitude();
+        const reset = SimVar.GetSimVarValue("L:A32NX_ALT_DEVIATION_SHORT", "Bool");
+        if (reset = true) {
```
Comment (vkrizan) [inline A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js:316]: This is an assignment to a constant. Either use comparison `===` or:
```suggestion
        if (reset) {
```

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js [hunk 5]
```diff
@@ -286,4 +309,81 @@ class A32NX_FWC {
...
+            this._wasInRange = false;
+            return;
+        }
+
+        if (this.warningPressed === true) {
+            this._wasBellowThreshold = false;
+            this._wasAboveThreshold = false;
+            this._wasInRange = false;
+            SimVar.SetSimVarValue("L:A32NX_ALT_DEVIATION", "Bool", false);
+            return;
```
Comment (vkrizan) [inline A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js:347]: Warning pressed condition should be moved before landing gear/glide slope conditions. Otherwise the continuous deviation sound cannot be inhibited by pressing the buttons if landing gear is down or glide slope is captured.

**Review by vkrizan — CHANGES_REQUESTED** [review]
Looks promising. There is one code issue...

Here are some bugs for the continuous warning:
1. cannot be canceled once gear is down (see inline comment)
1. master/caution warning not lit

And these are cases that might be checked with the manual for the continuous warning:
1. is not canceled when the alt. is changed
1. is not canceled when AP is engaged
1. is not canceled when reaching the selected alt.

**Comment by Kimbyeoungjang** [comment]
> Looks promising. There is one code issue...
> 
> Here are some bugs for the continuous warning:
> 
> 1. cannot be canceled once gear is down (see inline comment)
> 2. master/caution warning not lit
> 
> And these are cases that might be checked with the manual for the continuous warning:
> 
> 1. is not canceled when the alt. is changed
> 2. is not canceled when AP is engaged
> 3. is not canceled when reaching the selected alt.

oh i was not think about light lit. the caution light is lit when alarm on?

**Comment by vkrizan** [comment]
> and 3 is correct? in document, sound will canceled when press master warning or emer cle or new alt.

As stated in the reference, "The selection of a new altitude cancels the continuous C chord". This one seems to have been missed.

Clearing the continuous warning now works when gear is down. :+1:  

With the new changes I'm no longer hearing the short warning when approaching the sel. alt (AP is disengaged).

**Comment by Kimbyeoungjang** [comment]
> > and 3 is correct? in document, sound will canceled when press master warning or emer cle or new alt.
> 
> As stated in the reference, "The selection of a new altitude cancels the continuous C chord". This one seems to have been missed.
> 
> Clearing the continuous warning now works when gear is down. 👍
> 
> With the new changes I'm no longer hearing the short warning when approaching the sel. alt (AP is disengaged).

So current issue is short cchord sound? You can’t hear that when approching selected altitude?

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js [hunk 6]
```diff
@@ -65,6 +75,8 @@ class A32NX_FWC {
         this._updateButtons(_deltaTime);
         this._updateTakeoffMemo(_deltaTime);
         this._updateLandingMemo(_deltaTime);
+
+        this.altitudeWarning();
```
Comment (beheh) [inline A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js:79]:
```suggestion
        this._updatealtitudeWarning();
```
Just for naming consistency please.

File: .github/CHANGELOG.md [hunk 7]
```diff
@@ -75,6 +75,7 @@
 1. [SOUND] Fixed FCU knobs sounds - @ImenesFBW (Imenes)
 1. [MCDU] Fix AOC RCVD MSGS scrolling, add arrows, and update on receipt - @tracernz (Mike)
 1. [ECAM] Add TCAS STBY warning - @paul92ilm (Lussion)
+1. [System] Add Altitude Alert - @Kimbyeongjang (김병장#7165)
```
Comment (beheh) [inline .github/CHANGELOG.md:78]:
```suggestion
1. [FWC] Add aural Altitude Alert (C-Chord) - @Kimbyeongjang (김병장#7165)
```
I think this will help people recognize this better.

**Review by beheh — CHANGES_REQUESTED** [review]
One minor naming nitpick, otherwise this is good for now.

**Comment by vkrizan** [comment]
> So current issue is short cchord sound? You can’t hear that when approching selected altitude?

Yes, I did not hear it until my last comment. I haven't tried it yet, but I see there were some code changes that could have fixed that. The `reset === true` might have been the problem, as the simvar seems to return int for a Boolean value.

**Comment by Kimbyeoungjang** [comment]
> > So current issue is short cchord sound? You can’t hear that when approching selected altitude?
> 
> Yes, I did not hear it until my last comment. I haven't tried it yet, but I see there were some code changes that could have fixed that. The `reset === true` might have been the problem, as the simvar seems to return int for a Boolean value.

yes i fix that! now working bro

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js [hunk 8]
```diff
@@ -65,6 +75,7 @@ class A32NX_FWC {
         this._updateButtons(_deltaTime);
         this._updateTakeoffMemo(_deltaTime);
         this._updateLandingMemo(_deltaTime);
+        this._updatealtitudeWarning();
```
Comment (beheh) [inline A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js:78]: Sorry to nitpick again, but please follow camelCase:

```suggestion
        this._updateAltitudeWarning();
```

**Review by beheh — APPROVED** [review]
Looks good. There is a few subtleties that aren't encoded here like a hard inhibit while the aircraft is grounded, but these should have no practical impact here and will be modeled in the future FWC.

**Comment by beheh** [comment]
@Kimbyeoungjang Could you please add some testing instructions to your top comment? You could outline some scenarios to test so that the Q&A testers know how to replicate edge cases. For example you may tell them to flight to a specific altitude, and do some thing with the FCU and then leave that altitude, and that they should try the same with the AP and only observe the C Chord in specific scenarios.

**Comment by Kimbyeoungjang** [comment]
> @Kimbyeoungjang Could you please add some testing instructions to your top comment? You could outline some scenarios to test so that the Q&A testers know how to replicate edge cases. For example you may tell them to flight to a specific altitude, and do some thing with the FCU and then leave that altitude, and that they should try the same with the AP and only observe the C Chord in specific scenarios.

i added that :)

**Review by vkrizan — APPROVED** [review]

**Review by ilyeshammadi — APPROVED** [review]
Good work @Kimbyeoungjang 💯  can't wait to use it 😄

**Comment by Kimbyeoungjang** [comment]
oh i missied click sry beheh...

**Review by beheh — APPROVED** [review]
No problem, and thanks for adding the testing instructions. I've added Ready to Test and QA Tier 2 for now, as there do seem to be quite a few different edge cases that need testing.

**Comment by sidnov** [comment]
Will be testing it in the morning tomorrow if it doesn't get merged by then 👍🏻

**Comment by St54Kevin** [comment]
Category: QA Tester
Name: St54Kevin#3583
Date of testing: 09/02/2021
Version of the sim: 1.12.13.0
PR Tested: #3281 
Tier of testing: n/a
Changes to observe: ALT chord

Testing technique: 
Steps you've made to test the Pull Request:

1. Did a manual testflight and tested all situations I could find from the image below. 
2. no chords with AP on

Media:
image

Issues (If Any):
The chord sound is inconsistent and NOT as described as in the above picture. When you cancel an continous chord, it shortly stops and then it comes on again for 1.5s. After canceling the chord it comes on again even at a point where it shouldnt eg. at the 750ft mark. 

Overall Rating: Needs improvements to the chord

Conclusions of the testing: I advise @Kimbyeoungjang to testfly the reference picture in his PR. Just fly exactly the different situations like the arrows are pointing. Cancel the continuous chords to see what I mean. They shouldnt come on again at a point where it is not specified in the picture
The ALT pulsating display is correct in all cases! Good job.

**Comment by St54Kevin** [comment]
For all testers: It is mandatory to **push the ALT** knob everytime you change it in order for this PR to work. Otherwise its a mess

**Comment by Kimbyeoungjang** [comment]
> For all testers: It is mandatory to **push the ALT** knob everytime you change it in order for this PR to work. Otherwise its a mess

quick question.
idk in real aircraft, the selected altitude mean a FCU indicated altitude? or as my pr, always push alt knob to order aircraft climb or descend to indicated altitude?

and i'll change code quickly :)

**Comment by Kimbyeoungjang** [comment]
I solve that problem @St54Kevin
i wish this working :)

**Comment by St54Kevin** [comment]
> > For all testers: It is mandatory to **push the ALT** knob everytime you change it in order for this PR to work. Otherwise its a mess
> 
> quick question.
> idk in real aircraft, the selected altitude mean a FCU indicated altitude? or as my pr, always push alt knob to order aircraft climb or descend to indicated altitude?
> 
> and i'll change code quickly :)

I'm sorry but your English is very hard to understand :( The problem is with our A system. It only takes the new ALT when you push the knob no matter if the AP is on or off. IRL you don't have to push, turning the knob is enough there.

**Comment by Kimbyeoungjang** [comment]
> > > For all testers: It is mandatory to **push the ALT** knob everytime you change it in order for this PR to work. Otherwise its a mess
> > 
> > 
> > quick question.
> > idk in real aircraft, the selected altitude mean a FCU indicated altitude? or as my pr, always push alt knob to order aircraft climb or descend to indicated altitude?
> > and i'll change code quickly :)
> 
> I'm sorry but your English is very hard to understand :( The problem is with our AP system. It only takes the new ALT when you push the knob no matter if the AP is on or off. IRL you don't have to push, turning the knob is enough there.
> #2043 is the related issue

that's what i want to know.

so that is AP system problem and i don't have to change code until custom ap merge is it correct?

**Comment by Kimbyeoungjang** [comment]
@St54Kevin i fix that problem both PFD & FWC. Now working! you don't have to push or press altitude knob. just turn :)

**Comment by St54Kevin** [comment]
Still not perfect. I got a continous ALT chord right after takeoff and had to cancel it. Also I got a continous chord in the marked situations, when returning back to an 750ft mark AFTER leaving it
image

**Comment by St54Kevin** [comment]
Regarding the chord on takeoff, I think it takes the airport elev as a first ALT when you dont change the FCU ALT before takeoff. That should be fixed

**Comment by Kimbyeoungjang** [comment]
> Regarding the chord on takeoff, I think it takes the airport elev as a first ALT when you dont change the FCU ALT before takeoff. That should be fixed. It gets a flashing amber on takeoff and then a chord at +750ft

ok i fix that shortly. change to draft prevent unnecessary build. will update less than hour

**Comment by Kimbyeoungjang** [comment]
@St54Kevin sorry about my bad coding :(

now fix problems can you test it again?

**Comment by Kimbyeoungjang** [comment]
add default value.

**Comment by St54Kevin** [comment]
Still not correct. There is a 1,5s chord at the red markers, while it shouldnt according to the reference. Please fly the blue path and see.
image

**Comment by Kimbyeoungjang** [comment]
can you guys remove changes request & add ready to test?

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/PFD/AltimeterIndicator.js [hunk 9]
```diff
@@ -485,9 +485,10 @@ class Jet_PFD_AltimeterIndicator extends HTMLElement {
             return;
         }
 
-        const currentAltitudeConstraint = SimVar.GetSimVarValue("L:A32NX_AP_CSTN_ALT", "feet");
         // Use the constraint altitude if provided otherwise use selected altitude lock value
-        const targetAltitude = currentAltitudeConstraint && !this.getAutopilotMode() ? currentAltitudeConstraint : Simplane.getAutoPilotSelectedAltitudeLockValue();
+        const currentAltitudeConstraint = SimVar.GetSimVarValue("L:A32NX_AP_CSTN_ALT", "feet");
+        const currentFCUAltitude = SimVar.GetSimVarValue("L:HUD_AP_SELECTED_ALTITUDE", "Number");
+        const targetAltitude = currentAltitudeConstraint && !this.hasAltitudeConstraint() ? currentAltitudeConstraint : currentFCUAltitude;
```
Comment (vkrizan) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/PFD/AltimeterIndicator.js:491]: You can probably use `this.hudAPAltitude` replacing the SimVar get. :wink:

Comment (Kimbyeoungjang) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/PFD/AltimeterIndicator.js:491]: oh we can using that in there. I'ii change that.

**Comment by Kimbyeoungjang** [comment]
reduce simvar in pfd.

**Comment by Kimbyeoungjang** [comment]
> Takeoff is fixed. Chord is still not correct. There is a 1,5s chord at the red markers, while it shouldnt according to the reference (already stated this). Please fly the blue path, cancel the cont chord and see.
> image

if you don't mind, can you test this again?
all qa tester not working here :(

**Comment by marcman86** [comment]
Name: marcman86#4907
Date of testing: 14.2.2021
Version of the sim: 1.12.13.0
PR Tested: #3984
Tier of testing: 
Changes to observe: 
Testing technique: Tryout

Steps made to test the Pull Request:

1.) Spawned on Runway at EDDP
2.) Set managed alt to 7000FT and heading
3.) Took off an turn on AP to get to 7000FT

With AP no sound should appear (works).

4.) Switchted to manual flight. Flew all possibilities of the reference image.

5.) Looked for right flashing color and sounds (works)

Issues (minor things): 
- Seems the linewidth of the amber altitude window is smaller than from the yellow pulsing.
- Should there be a difference between flashing and pulsing?

Overall Rating: Good

Conclusions of the testing: Core works. Just minor details to fix.

Hope this helps :-)

**Comment by sidnov** [comment]
@marcman86 here's a good reference for the flashing :) https://youtu.be/TEi-WX08ov8?t=122

**Comment by beheh** [comment]
There seems to be some kind of confusion here. The flashing of the altitude window has already been implemented and released a long, long time ago (already in stable). As such, it is unrelated to this PR and there is no need to test it (although it should be working correctly).

What this PR *does* do is introduce the C Chord - a warning sound that is emitted by the aircraft with similar logic to the flashing, however like other warning sounds it can be cancelled. That is the part that needs to be tested according to the diagram above.

**Comment by sidnov** [comment]
> There seems to be some kind of confusion here. The flashing of the altitude window has already been implemented and released a long, long time ago (already in stable). As such, it is unrelated to this PR and there is no need to test it (although it should be working correctly).
> 
> What this PR _does_ do is introduce the C Chord - a warning sound that is emitted by the aircraft with similar logic to the flashing, however like other warning sounds it can be cancelled. That is the part that needs to be tested according to the diagram above.

Thanks for the clarification! I am currently testing it as I write this comment. One thing I noticed so far is, although the C-chord is classified as a warning sound, it is not linked to the right soundset within the sim. I found this by muting "Warning" sounds in the settings panel, which disabled AP disconnect sound, overspeed warning sound, but not the C-Chord.

**Comment by marcman86** [comment]
> 
> 
> @marcman86 here's a good reference for the flashing :) https://youtu.be/TEi-WX08ov8?t=122

This video only shows the yellow pulsing. I think there is a difference between pulsing and flashing. Cause why do they name it differently?

When this PR only handles the sound we have to look at this another time. Sound works fine in the test.

**Comment by sidnov** [comment]
Category: QA Tester
Name: sidnov#8337
Date of testing: 16.2.2021
Version of the sim: 1.12.13.0
PR Tested: #3281 
Tier of testing: 2

### Changes to observe: Altitude C-Chord warning

### Testing technique: Tryout

### Steps you've made to test the Pull Request:

1. Spawned on runway at VABB
2. Set altitude on FCU to 4000ft
3. Departed RWY27 on runway heading
4. Climbed on AP to 4000ft, ob
5. Climbed further to 5000ft on AP, descended to 4000ft again on AP, observed no warning
6. Performed the maneuvers required to trigger the warnings passing -750, -250, +250, +750 on both ascending and descending, along with the intermediate conditions.

### Issues (If Any): As mentioned in a comment above, although the C-chord is classified as a warning sound, it is not linked to the right soundset within the sim. I found this by muting "Warning" sounds in the settings panel, which disabled AP disconnect sound, overspeed warning sound, but not the C-Chord. 

### Overall Rating: Good

### Conclusions of the testing: All the conditions were tested and the warning sounded as expected at each point. Good to go!

**Comment by kberg93** [comment]
Category: QA Tester
Name: Kberg#0666
Date of testing: Current date 02/19/2021
Version of the sim: 1.13.16.0
PR Tested: #3281 
Tier of testing: 2
Changes to observe:
Added altitude C chord warning sounds

Testing technique: Flying manually testing alerts.

Steps you've made to test the Pull Request: 
1. Spawn at ENVA rwy00 take off and climb to 7000ft with ap.
2. No warnings on changing alt with AP.
3. Disengaged AP and performed various of the tests on the picture above, all went well.
4. Landing at ENVA rwy09 testing that warnings ar inhibited when slats and ldg is out and it worked as descirbed.

Flight notes (When done a full flight): ENVA09 to ENVA09 manual test flight.

Media: Insert screenshots, recordings, or anything relevant to the review here.

Issues (If Any): None.

Overall Rating: Good.
Conclusions of the testing: All of the functions worked as described.

**Review by beheh — APPROVED** [review]

**Comment by ChruutvoLuzi** [comment]
Category: QA Tester
Name: ChruutvoLuzi | SD-PB.ch#8902
Date of testing: 20.2.2021
Version of the sim: 1.13.16.0
PR Tested: #3281 
Tier of testing: 2
Changes to observe: CChord Altitude Warning


Testing technique: TryOut
Steps you've made to test the Pull Request:
1. Started C&D at EGNT
2. Started up the Aircraft
3. Took off
4. selected 5000ft and climbed by hand = checked c chord between 4250 and 4750ft and yellow flashing
5. deviated 1000ft to 4000ft = checked for amber flashing and continous c chord
6. climbed back to 5000ft by hand = yellow flashing by 4250 and no cchord by 4750 (see issue 1)
7. climbed to 6000ft using AP = checked for no cchord and yellow flashing
8. disconnected AP and  climbed by hand to 6500 = checked for continous cchord and amber flashing
9. descended to 6000ft again = checked cchord and flasing stopping by 6250ft (see issue 1)
10. selected 8000ft and started climb by hand, stopped climb by yellow flashing, descended and checked for amber flashing passing 7250ft
11. climbed back to 8000ft = checked for yellow flashing when passing 7250ft and no cchord (see issue 1)

Media:
https://cdn.discordapp.com/attachments/314015589595086848/812758653782327296/Microsoft_Flight_Simulator_2021.02.20_-_19.50.22.02.mp4

Issues (If Any):
- continous cchord does not cancel after climbing/descending back to selected Altitude
Steps to achieve the issue
select 5000ft
climb to 4500ft
descend to 4000ft to get continous cchord and amber flashing
climb to selected 5000ft 

Overall Rating: To be improved
Conclusions of the testing: As you can see and hear, the continous cchord does not cancel. When climbing back to the selected Altitude, you can hear the other cchord overlapping the continous cchord at the +/- 750 ft, but again it doesnt cancel

**Comment by Kimbyeoungjang** [comment]
oh god... there was no issue 1 situation on fwc document XDDD
I can fix that issue 1 & 2. but can you give me details about issue 3? @ChruutvoLuzi 

Cont Chord should not working when landing(mean when gear is down or slats extended).
I need more info about issue 3 the you testing.

**Comment by Kimbyeoungjang** [comment]
change to draft half hour to prevent github check

**Comment by Kimbyeoungjang** [comment]
I fix init error but wtf is that other error. I never touch EFB things wtf

**Comment by beheh** [comment]
Can we get one more QA test for this after the last changes?

**Comment by marcman86** [comment]
Category: QA Tester
Name: marcman86#4907
Date of testing: 23.2.2021
Version of the sim: 1.13.16.0
PR Tested: #3281
Tier of testing: 2
Changes to observe: C-chords
Testing technique: Tryout

Steps made to test the Pull Request:

1.) Spawned runway at EDDP
2.) Took off to 5000ft
3.) Tested flight pattern
4.) Tested inhibit: 
- Slats out and gear selected down 
- Gear down locked
- Manual G/S

Issues: -

Overall Rating: Good

Conclusions of the testing: Works

Hope this helps :-)

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js [hunk 10]
```diff
@@ -89,6 +101,17 @@ class A32NX_FWC {
...
+        if (SimVar.GetSimVarValue("L:PUSH_AUTOPILOT_MASTERAWARN_L", "Bool") || SimVar.GetSimVarValue("L:PUSH_AUTOPILOT_MASTERAWARN_R", "Bool")) {
+            this.warningPressed = true;
+        } else {
+            this.warningPressed = false;
+        }
+        if (SimVar.GetSimVarValue("L:PUSH_AUTOPILOT_MASTERCAUT_L", "Bool") || SimVar.GetSimVarValue("L:PUSH_AUTOPILOT_MASTERCAUT_R", "Bool")) {
+            this.cautionPressed = true;
+        } else {
+            this.cautionPressed = false;
+        }
```
Comment (davidwalschots) [inline A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js:114]: **question**: What was the intent of this code? The variables are set but never used.

File: A32NX/SimObjects/AirPlanes/Asobo_A320_NEO/model/A320_NEO_INTERIOR.xml [hunk 11]
```diff
@@ -3419,6 +3419,25 @@
...
+            </UseTemplate>
+            <UseTemplate Name="FBW_Push_Held">
+                <NODE_ID>PUSH_AUTOPILOT_MASTERCAUT_L</NODE_ID>
+                <HOLD_SIMVAR>L:PUSH_AUTOPILOT_MASTERCAUT_L</HOLD_SIMVAR>
+            </UseTemplate>
+            <UseTemplate Name="FBW_Push_Held">
+                <NODE_ID>PUSH_AUTOPILOT_MASTERCAUT_R</NODE_ID>
+                <HOLD_SIMVAR>L:PUSH_AUTOPILOT_MASTERCAUT_R</HOLD_SIMVAR>
+            </UseTemplate>
+        </Component>
```
Comment (davidwalschots) [inline A32NX/SimObjects/AirPlanes/Asobo_A320_NEO/model/A320_NEO_INTERIOR.xml:3439]: Adding this broke the ability to press the MASTER WARN and MASTER CAUT button to stop sounds. E.g. when you press the throttle fully forward with the parking brake set, you cannot stop the audio by pressing the MASTER WARN button anymore. The NODE_IDs are already used above. They're already assigned.

File: A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js (hunk 10 above)
Comment (Kimbyeoungjang) [reply A32NX/html_ui/Pages/A32NX_Core/A32NX_FWC.js:114]: because on future, I want to control caution on FWC

File: A32NX/SimObjects/AirPlanes/Asobo_A320_NEO/model/A320_NEO_INTERIOR.xml (hunk 11 above)
Comment (Kimbyeoungjang) [reply A32NX/SimObjects/AirPlanes/Asobo_A320_NEO/model/A320_NEO_INTERIOR.xml:3439]: I want to add that function on FWC because some warning should not canceled by master warning and caution.
